src/NServiceBus.Extensions.Hosting/HostAwareMessageSession.cs

```diff
@@ -0,0 +1,73 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+
+    class HostAwareMessageSession : IMessageSession
+    {
+        public HostAwareMessageSession(Lazy<IMessageSession> messageSession)
+        {
+            this.messageSession = messageSession;
+        }
+
+        public Task Send(object message, SendOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Send(message, options);
+        }
+
+        public Task Send<T>(Action<T> messageConstructor, SendOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Send(messageConstructor, options);
+        }
+
+        public Task Publish(object message, PublishOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Publish(message, options);
+        }
+
+        public Task Publish<T>(Action<T> messageConstructor, PublishOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Publish(messageConstructor, options);
+        }
+
+        public Task Subscribe(Type eventType, SubscribeOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Subscribe(eventType, options);
+        }
+
+        public Task Unsubscribe(Type eventType, UnsubscribeOptions options)
+        {
+            GuardAgainstTooEarlyUse();
+
+            return messageSession.Value.Unsubscribe(eventType, options);
+        }
+
+        public void MarkReadyForUse()
+        {
+            isReadyForUse = true;
```

Comment:
Do you see any problems because we are not using a thread safe way to toggle this?

Reply:
I couldn't come up with a scenario where this would be an issue assuming that the host will call all the .StartAsync methods sequentially and in order?

Reply:
I was thinking about the toggle of the bool not being seen/read in the new state by web api request that run concurrently. Probably a non-issue though